In [ ]:
#define objective------
#import all lib-----
#load the data-----
#data analysis------
#Data preprocessing---------
#split data into train test split (train,test)#
#selet algorithm
#train your model
#test your model
#check accuracy (80%)
#prediction/classification

In [1]:
import numpy as np 
import pandas as pd #data manipulation tool
import matplotlib.pyplot as plt

In [3]:
from datetime import datetime

# Train Data:

In [2]:
house_data=pd.read_csv("Predicting-House-Prices-In-Bengaluru-Train-Data.csv")

In [7]:
house_data.head(10)

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   
5  Super built-up  Area  Ready To Move                Whitefield      2 BHK   
6  Super built-up  Area         18-May          Old Airport Road      4 BHK   
7  Super built-up  Area  Ready To Move              Rajaji Nagar      4 BHK   
8  Super built-up  Area  Ready To Move              Marathahalli      3 BHK   
9            Plot  Area  Ready To Move              Gandhi Bazar  6 Bedroom   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00  
5  DuenaTa       1170   2.0      1.0   38.00  
6  Jaades        2732   4.0      NaN  204.00  
7  Brway G       3300   4.0      NaN  600.00  
8      NaN       1310   3.0      1.0   63.25  
9      NaN       1020   6.0      NaN  370.00

In [11]:
house_data.isnull().sum() #finding sum of null values

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

In [12]:
house_data.describe() #know all stats in your dataset

bath       balcony         price
count  13247.000000  12711.000000  13320.000000
mean       2.692610      1.584376    112.565627
std        1.341458      0.817263    148.971674
min        1.000000      0.000000      8.000000
25%        2.000000      1.000000     50.000000
50%        2.000000      2.000000     72.000000
75%        3.000000      2.000000    120.000000
max       40.000000      3.000000   3600.000000

In [8]:
house_data.info() #retrive all data type info 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [11]:
house_data.nunique()

area_type          4
availability      81
location        1305
size              31
society         2688
total_sqft      2117
bath              19
balcony            4
price           1994
dtype: int64

## No-of-missing values in soceity = 5502 {huge! We Remove it..}
## location has more no of unique values,hence it would not affect the house prices {Remove!}

In [15]:
house_data.drop('society',axis=1,inplace=True)
house_data.drop('location',axis=1,inplace=True)

In [18]:
house_data.head(10)

area_type   availability       size total_sqft  bath  balcony  \
0  Super built-up  Area         19-Dec      2 BHK       1056   2.0      1.0   
1            Plot  Area  Ready To Move  4 Bedroom       2600   5.0      3.0   
2        Built-up  Area  Ready To Move      3 BHK       1440   2.0      3.0   
3  Super built-up  Area  Ready To Move      3 BHK       1521   3.0      1.0   
4  Super built-up  Area  Ready To Move      2 BHK       1200   2.0      1.0   
5  Super built-up  Area  Ready To Move      2 BHK       1170   2.0      1.0   
6  Super built-up  Area         18-May      4 BHK       2732   4.0      NaN   
7  Super built-up  Area  Ready To Move      4 BHK       3300   4.0      NaN   
8  Super built-up  Area  Ready To Move      3 BHK       1310   3.0      1.0   
9            Plot  Area  Ready To Move  6 Bedroom       1020   6.0      NaN   

    price  
0   39.07  
1  120.00  
2   62.00  
3   95.00  
4   51.00  
5   38.00  
6  204.00  
7  600.00  
8   63.25  
9  370.00

# Dummy variables

In [19]:
area_type_dummies = pd.get_dummies(house_data['area_type'],drop_first=True)
area_type_dummies

Carpet  Area  Plot  Area  Super built-up  Area
0                 0           0                     1
1                 0           1                     0
2                 0           0                     0
3                 0           0                     1
4                 0           0                     1
...             ...         ...                   ...
13315             0           0                     0
13316             0           0                     1
13317             0           0                     0
13318             0           0                     1
13319             0           0                     1

[13320 rows x 3 columns]

In [20]:
house_data.drop('area_type',axis=1,inplace=True) 

In [21]:
house_data=pd.concat([house_data,area_type_dummies],axis=1)

In [22]:
house_data.head()

availability       size total_sqft  bath  balcony   price  Carpet  Area  \
0         19-Dec      2 BHK       1056   2.0      1.0   39.07             0   
1  Ready To Move  4 Bedroom       2600   5.0      3.0  120.00             0   
2  Ready To Move      3 BHK       1440   2.0      3.0   62.00             0   
3  Ready To Move      3 BHK       1521   3.0      1.0   95.00             0   
4  Ready To Move      2 BHK       1200   2.0      1.0   51.00             0   

   Plot  Area  Super built-up  Area  
0           0                     1  
1           1                     0  
2           0                     0  
3           0                     1  
4           0                     1

### Missing values

In [24]:
house_data['bath'].mean()

2.6926096474673513

In [26]:
house_data['bath']=house_data['bath'].fillna(value=house_data['bath'].mean())

In [34]:
#house_data.info()

In [31]:
house_data['balcony'].fillna(value=house_data['balcony'].mean(),inplace=True)

In [32]:
house_data.head()

availability       size total_sqft  bath  balcony   price  Carpet  Area  \
0         19-Dec      2 BHK       1056   2.0      1.0   39.07             0   
1  Ready To Move  4 Bedroom       2600   5.0      3.0  120.00             0   
2  Ready To Move      3 BHK       1440   2.0      3.0   62.00             0   
3  Ready To Move      3 BHK       1521   3.0      1.0   95.00             0   
4  Ready To Move      2 BHK       1200   2.0      1.0   51.00             0   

   Plot  Area  Super built-up  Area  
0           0                     1  
1           1                     0  
2           0                     0  
3           0                     1  
4           0                     1

In [35]:
pd.unique(house_data['size'])

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', nan, '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

### Size couloumns-> numbers

In [36]:
#create emplty list
size_bhk = []

#import regular expression
import re

pattern = r"BHK"
for i in house_data['size']:
    if re.search(pattern,str(i)):
       
        size_bhk.append(int(i[0]))
    else:
        size_bhk.append(0)    
        
size_bed = []   

pattern = r"Bedroom"
for i in house_data['size']:
    if re.search(pattern, str(i)):
      
        size_bed.append(int(i[0]))
    else:
        size_bed.append(0)

In [41]:
size_bhk

[2,
 0,
 3,
 3,
 2,
 2,
 4,
 4,
 3,
 0,
 3,
 0,
 2,
 2,
 0,
 2,
 3,
 3,
 3,
 2,
 1,
 3,
 0,
 3,
 0,
 3,
 2,
 3,
 2,
 3,
 4,
 3,
 3,
 3,
 3,
 2,
 2,
 3,
 3,
 2,
 2,
 3,
 1,
 0,
 2,
 0,
 2,
 2,
 0,
 2,
 2,
 3,
 3,
 2,
 3,
 2,
 0,
 2,
 0,
 2,
 0,
 2,
 0,
 2,
 0,
 2,
 2,
 2,
 0,
 2,
 0,
 2,
 3,
 2,
 3,
 2,
 2,
 3,
 2,
 0,
 3,
 0,
 2,
 3,
 3,
 0,
 3,
 2,
 2,
 0,
 2,
 2,
 3,
 0,
 2,
 3,
 0,
 3,
 2,
 0,
 3,
 2,
 0,
 2,
 2,
 2,
 3,
 3,
 2,
 2,
 3,
 3,
 2,
 3,
 0,
 0,
 3,
 3,
 2,
 0,
 3,
 3,
 4,
 4,
 3,
 2,
 3,
 2,
 2,
 0,
 3,
 0,
 2,
 0,
 3,
 2,
 2,
 2,
 3,
 3,
 0,
 5,
 3,
 2,
 2,
 7,
 3,
 2,
 3,
 0,
 3,
 2,
 3,
 3,
 2,
 2,
 2,
 3,
 3,
 0,
 3,
 2,
 3,
 0,
 3,
 2,
 3,
 3,
 2,
 3,
 6,
 1,
 2,
 3,
 3,
 0,
 0,
 0,
 1,
 2,
 2,
 3,
 3,
 3,
 2,
 0,
 2,
 2,
 2,
 2,
 0,
 3,
 0,
 0,
 2,
 2,
 2,
 0,
 3,
 2,
 2,
 3,
 2,
 2,
 3,
 3,
 3,
 2,
 2,
 0,
 4,
 3,
 3,
 3,
 3,
 2,
 0,
 1,
 3,
 3,
 3,
 3,
 4,
 2,
 3,
 3,
 2,
 0,
 3,
 2,
 2,
 2,
 2,
 2,
 2,
 3,
 2,
 3,
 0,
 2,
 1,
 0,
 1,
 3,
 2,
 0,
 2,
 3,
 0,
 3,


In [42]:
size_bed

[0,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 6,
 0,
 4,
 0,
 0,
 3,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 8,
 0,
 0,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 4,
 0,
 6,
 0,
 3,
 0,
 4,
 0,
 8,
 0,
 0,
 0,
 8,
 0,
 3,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 6,
 0,
 4,
 0,
 0,
 0,
 6,
 0,
 0,
 0,
 6,
 0,
 0,
 0,
 4,
 0,
 0,
 4,
 0,
 0,
 3,
 0,
 0,
 3,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 4,
 3,
 0,
 0,
 0,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 7,
 0,
 3,
 0,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 8,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 6,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 4,
 0,
 0,
 0,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 3,
 5,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 3,
 0,
 0,
 0,
 0,
 4,
 0,
 4,
 7,
 0,
 0,
 0,
 3,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 6,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 5,
 0,
 0,
 8,
 0,
 0,
 0,
 3,
 0,
 0,
 3,
 0,


In [43]:
house_data['BHK'] = pd.Series(size_bhk,index=house_data.index)
house_data['Bedroom'] = pd.Series(size_bed,index=house_data.index)

In [44]:
house_data.head(10)

availability       size total_sqft  bath   balcony   price  Carpet  Area  \
0         19-Dec      2 BHK       1056   2.0  1.000000   39.07             0   
1  Ready To Move  4 Bedroom       2600   5.0  3.000000  120.00             0   
2  Ready To Move      3 BHK       1440   2.0  3.000000   62.00             0   
3  Ready To Move      3 BHK       1521   3.0  1.000000   95.00             0   
4  Ready To Move      2 BHK       1200   2.0  1.000000   51.00             0   
5  Ready To Move      2 BHK       1170   2.0  1.000000   38.00             0   
6         18-May      4 BHK       2732   4.0  1.584376  204.00             0   
7  Ready To Move      4 BHK       3300   4.0  1.584376  600.00             0   
8  Ready To Move      3 BHK       1310   3.0  1.000000   63.25             0   
9  Ready To Move  6 Bedroom       1020   6.0  1.584376  370.00             0   

   Plot  Area  Super built-up  Area  BHK  Bedroom  
0           0                     1    2        0  
1           1                     0    0        4  
2           0                     0    3        0  
3           0                     1    3        0  
4           0                     1    2        0  
5           0                     1    2        0  
6           0                     1    4        0  
7           0                     1    4        0  
8           0                     1    3        0  
9           1                     0    0        6

In [45]:
house_data.drop('size',axis=1,inplace=True)

In [46]:
house_data

availability total_sqft  bath   balcony   price  Carpet  Area  \
0             19-Dec       1056   2.0  1.000000   39.07             0   
1      Ready To Move       2600   5.0  3.000000  120.00             0   
2      Ready To Move       1440   2.0  3.000000   62.00             0   
3      Ready To Move       1521   3.0  1.000000   95.00             0   
4      Ready To Move       1200   2.0  1.000000   51.00             0   
...              ...        ...   ...       ...     ...           ...   
13315  Ready To Move       3453   4.0  0.000000  231.00             0   
13316  Ready To Move       3600   5.0  1.584376  400.00             0   
13317  Ready To Move       1141   2.0  1.000000   60.00             0   
13318         18-Jun       4689   4.0  1.000000  488.00             0   
13319  Ready To Move        550   1.0  1.000000   17.00             0   

       Plot  Area  Super built-up  Area  BHK  Bedroom  
0               0                     1    2        0  
1               1                     0    0        4  
2               0                     0    3        0  
3               0                     1    3        0  
4               0                     1    2        0  
...           ...                   ...  ...      ...  
13315           0                     0    0        5  
13316           0                     1    4        0  
13317           0                     0    2        0  
13318           0                     1    4        0  
13319           0                     1    1        0  

[13320 rows x 10 columns]

### Date time series:

#### Reading availability date:

In [48]:
#empty list
avail_date = []

for item in house_data['availability']:
    if(item[0].isdigit()):
        item = str(datetime.today().date().year) + '-' + item
        avail_date.append((datetime.strptime(item , '%Y-%d-%b')).date())
    else:
        avail_date.append(0)

In [49]:
avail_date

[datetime.date(2021, 12, 19),
 0,
 0,
 0,
 0,
 0,
 datetime.date(2021, 5, 18),
 0,
 0,
 0,
 datetime.date(2021, 2, 18),
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 datetime.date(2021, 12, 19),
 0,
 0,
 datetime.date(2021, 11, 18),
 datetime.date(2021, 5, 18),
 0,
 datetime.date(2021, 12, 20),
 datetime.date(2021, 10, 17),
 0,
 datetime.date(2021, 12, 19),
 0,
 0,
 0,
 0,
 datetime.date(2021, 12, 21),
 0,
 0,
 0,
 0,
 0,
 datetime.date(2021, 12, 19),
 0,
 0,
 datetime.date(2021, 9, 19),
 0,
 0,
 datetime.date(2021, 9, 20),
 0,
 0,
 0,
 datetime.date(2021, 3, 18),
 0,
 0,
 0,
 0,
 datetime.date(2021, 2, 20),
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 datetime.date(2021, 4, 18),
 0,
 0,
 0,
 0,
 datetime.date(2021, 8, 20),
 0,
 datetime.date(2021, 4, 18),
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 datetime.date(2021, 10, 18),
 datetime.date(2021, 3, 19),
 0,
 0,
 0,
 0,
 datetime.date(2021, 9, 17),
 0,
 0,
 0,
 datetime.date(2021, 12, 18),
 0,
 0,
 datetime.date(2021, 8, 17),
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,

#### no of days till availalability from today

In [50]:

days_till_available = []

for item in avail_date:
    if(item != 0):
        n_days = item - datetime.today().date()
        if(n_days.days > 0):
            days_till_available.append(n_days.days)
        else:
            days_till_available.append(n_days.days + 365)
    else:
        days_till_available.append(0)

In [51]:
days_till_available

[309,
 0,
 0,
 0,
 0,
 0,
 94,
 0,
 0,
 0,
 5,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 309,
 0,
 0,
 278,
 94,
 0,
 310,
 246,
 0,
 309,
 0,
 0,
 0,
 0,
 311,
 0,
 0,
 0,
 0,
 0,
 309,
 0,
 0,
 218,
 0,
 0,
 219,
 0,
 0,
 0,
 33,
 0,
 0,
 0,
 0,
 7,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 64,
 0,
 0,
 0,
 0,
 188,
 0,
 64,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 247,
 34,
 0,
 0,
 0,
 0,
 216,
 0,
 0,
 0,
 308,
 0,
 0,
 185,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 65,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 308,
 125,
 0,
 0,
 0,
 33,
 0,
 312,
 0,
 0,
 343,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 186,
 34,
 94,
 340,
 0,
 309,
 0,
 0,
 154,
 0,
 0,
 0,
 0,
 0,
 65,
 0,
 94,
 0,
 0,
 155,
 0,
 0,
 0,
 0,
 0,
 94,
 0,
 0,
 0,
 308,
 0,
 0,
 0,
 0,
 0,
 0,
 64,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 128,
 0,
 0,
 0,
 0,
 0,
 0,
 308,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 96,
 0,
 125,
 0,
 0,
 0,
 0,
 0,
 0,
 308,
 0,
 0,
 0,
 34,
 309,
 0,
 0,
 0,
 0,
 187,
 0

In [52]:
house_data['days_till_available'] = pd.Series(days_till_available , index=house_data.index)
house_data.head()

availability total_sqft  bath  balcony   price  Carpet  Area  Plot  Area  \
0         19-Dec       1056   2.0      1.0   39.07             0           0   
1  Ready To Move       2600   5.0      3.0  120.00             0           1   
2  Ready To Move       1440   2.0      3.0   62.00             0           0   
3  Ready To Move       1521   3.0      1.0   95.00             0           0   
4  Ready To Move       1200   2.0      1.0   51.00             0           0   

   Super built-up  Area  BHK  Bedroom  days_till_available  
0                     1    2        0                  309  
1                     0    0        4                    0  
2                     0    3        0                    0  
3                     1    3        0                    0  
4                     1    2        0                    0

In [53]:
house_data.drop('availability',axis=1,inplace=True)

In [54]:
house_data.head(10)

total_sqft  bath   balcony   price  Carpet  Area  Plot  Area  \
0       1056   2.0  1.000000   39.07             0           0   
1       2600   5.0  3.000000  120.00             0           1   
2       1440   2.0  3.000000   62.00             0           0   
3       1521   3.0  1.000000   95.00             0           0   
4       1200   2.0  1.000000   51.00             0           0   
5       1170   2.0  1.000000   38.00             0           0   
6       2732   4.0  1.584376  204.00             0           0   
7       3300   4.0  1.584376  600.00             0           0   
8       1310   3.0  1.000000   63.25             0           0   
9       1020   6.0  1.584376  370.00             0           1   

   Super built-up  Area  BHK  Bedroom  days_till_available  
0                     1    2        0                  309  
1                     0    0        4                    0  
2                     0    3        0                    0  
3                     1    3        0                    0  
4                     1    2        0                    0  
5                     1    2        0                    0  
6                     1    4        0                   94  
7                     1    4        0                    0  
8                     1    3        0                    0  
9                     0    0        6                    0

In [55]:
house_data.dtypes

total_sqft               object
bath                    float64
balcony                 float64
price                   float64
Carpet  Area              uint8
Plot  Area                uint8
Super built-up  Area      uint8
BHK                       int64
Bedroom                   int64
days_till_available       int64
dtype: object

In [62]:
a=[1,2,3,4,5]
a[0:-2]

[1, 2, 3]

### converting to sq feets

In [63]:
TOTAL_sqft = []

for item in house_data['total_sqft']:
    if('-' in str(item)):
        temp = item.split()
        temp1 = (float(temp[0]) + float(temp[2]))/2  #for values like(125-500)
        TOTAL_sqft.append(temp1)
    elif(item.endswith("Sq. Meter")):       #converting to sq feets
        temp = float(item[:-9])              #excluding last 9
        TOTAL_sqft.append(temp * 10.7639)    
    elif(item.endswith("Perch")):
        temp = float(item[:-5])
        TOTAL_sqft.append(temp * 272.25)
    elif(item.endswith("Acres")):
        temp = float(item[:-5])
        TOTAL_sqft.append(temp * 43560)
    elif(item.endswith("Sq. Yards")):
        temp = float(item[:-9])
        TOTAL_sqft.append(temp * 9)
    elif(item.endswith("Cents")):
        temp = float(item[:-5])
        TOTAL_sqft.append(temp * 435.6)
    elif(item.endswith("Guntha")):
        temp = float(item[:-6])
        TOTAL_sqft.append(temp * 1089)
    elif(item.endswith("Grounds")):
        temp = float(item[:-7])
        TOTAL_sqft.append(temp * 2400)
    else:
        TOTAL_sqft.append(float(item))

In [64]:
TOTAL_sqft

[1056.0,
 2600.0,
 1440.0,
 1521.0,
 1200.0,
 1170.0,
 2732.0,
 3300.0,
 1310.0,
 1020.0,
 1800.0,
 2785.0,
 1000.0,
 1100.0,
 2250.0,
 1175.0,
 1180.0,
 1540.0,
 2770.0,
 1100.0,
 600.0,
 1755.0,
 2800.0,
 1767.0,
 510.0,
 1250.0,
 660.0,
 1610.0,
 1151.0,
 1025.0,
 2475.0,
 1075.0,
 1760.0,
 1693.0,
 1925.0,
 700.0,
 1070.0,
 1724.0,
 1290.0,
 1143.0,
 1296.0,
 1254.0,
 600.0,
 660.0,
 1330.74,
 600.0,
 970.0,
 1459.0,
 800.0,
 869.0,
 1270.0,
 1670.0,
 2010.0,
 1185.0,
 1600.0,
 1200.0,
 3210.0,
 1500.0,
 1407.0,
 840.0,
 4395.0,
 845.0,
 5700.0,
 1160.0,
 3000.0,
 1100.0,
 1140.0,
 1220.0,
 1350.0,
 1005.0,
 500.0,
 1358.0,
 1569.0,
 1240.0,
 2089.0,
 1206.0,
 1150.0,
 2511.0,
 460.0,
 4400.0,
 1660.0,
 3203.5,
 1326.0,
 1325.0,
 1499.0,
 3000.0,
 1665.0,
 708.0,
 1060.0,
 710.0,
 1000.0,
 1000.0,
 1450.0,
 1200.0,
 1296.0,
 1540.0,
 2894.0,
 1330.0,
 1200.0,
 1200.0,
 2502.0,
 650.0,
 2400.0,
 1007.0,
 1200.0,
 966.0,
 1630.0,
 1640.0,
 782.0,
 1260.0,
 1800.0,
 1413.0,
 1116.0,
 

In [65]:
house_data['TOTAL_sqft'] = pd.Series(TOTAL_sqft, index=house_data.index)

In [66]:
house_data.drop(columns='total_sqft',inplace=True)

In [67]:
house_data.head()

bath  balcony   price  Carpet  Area  Plot  Area  Super built-up  Area  BHK  \
0   2.0      1.0   39.07             0           0                     1    2   
1   5.0      3.0  120.00             0           1                     0    0   
2   2.0      3.0   62.00             0           0                     0    3   
3   3.0      1.0   95.00             0           0                     1    3   
4   2.0      1.0   51.00             0           0                     1    2   

   Bedroom  days_till_available  TOTAL_sqft  
0        0                  309      1056.0  
1        4                    0      2600.0  
2        0                    0      1440.0  
3        0                    0      1521.0  
4        0                    0      1200.0

## TEST DATA

In [69]:
house_test_data = pd.read_csv('Predicting-House-Prices-In-Bengaluru-Test-Data.csv')

In [70]:
house_test_data.head()

area_type   availability           location       size  society  \
0  Super built-up  Area  Ready To Move        Brookefield      2 BHK  Roeekbl   
1            Plot  Area  Ready To Move      Akshaya Nagar  9 Bedroom      NaN   
2            Plot  Area         18-Apr        Hennur Road  4 Bedroom  Saandtt   
3  Super built-up  Area  Ready To Move  Kodichikkanahalli      3 BHK  Winerri   
4  Super built-up  Area  Ready To Move        Konanakunte      2 BHK  AmageSa   

  total_sqft  bath  balcony  price  
0       1225   2.0      2.0    NaN  
1       2400   9.0      2.0    NaN  
2       1650   5.0      2.0    NaN  
3       1322   3.0      1.0    NaN  
4       1161   2.0      1.0    NaN

In [71]:
house_test_data.drop('society',axis=1,inplace=True)
house_test_data.drop('location',axis=1,inplace=True)
house_test_data.head()

area_type   availability       size total_sqft  bath  balcony  \
0  Super built-up  Area  Ready To Move      2 BHK       1225   2.0      2.0   
1            Plot  Area  Ready To Move  9 Bedroom       2400   9.0      2.0   
2            Plot  Area         18-Apr  4 Bedroom       1650   5.0      2.0   
3  Super built-up  Area  Ready To Move      3 BHK       1322   3.0      1.0   
4  Super built-up  Area  Ready To Move      2 BHK       1161   2.0      1.0   

   price  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

In [72]:
area_type_dummies = pd.get_dummies(house_test_data['area_type'],drop_first=True)
house_test_data=pd.concat([house_test_data,area_type_dummies],axis=1)
house_test_data.head()

area_type   availability       size total_sqft  bath  balcony  \
0  Super built-up  Area  Ready To Move      2 BHK       1225   2.0      2.0   
1            Plot  Area  Ready To Move  9 Bedroom       2400   9.0      2.0   
2            Plot  Area         18-Apr  4 Bedroom       1650   5.0      2.0   
3  Super built-up  Area  Ready To Move      3 BHK       1322   3.0      1.0   
4  Super built-up  Area  Ready To Move      2 BHK       1161   2.0      1.0   

   price  Carpet  Area  Plot  Area  Super built-up  Area  
0    NaN             0           0                     1  
1    NaN             0           1                     0  
2    NaN             0           1                     0  
3    NaN             0           0                     1  
4    NaN             0           0                     1

In [73]:
house_test_data.drop('area_type',axis=1,inplace=True)

In [74]:
size_bhk = []
import re
pattern = r"BHK"
for i in house_test_data['size']:
    if re.search(pattern,str(i)):
       
        size_bhk.append(int(i[0]))
    else:
        size_bhk.append(0)

        
                
size_bed = []       
pattern = r"Bedroom"
for i in house_test_data['size']:
    if re.search(pattern, str(i)):
      
        size_bed.append(int(i[0]))
    else:
        size_bed.append(0)

house_test_data['BHK'] = pd.Series(size_bhk,index=house_test_data.index)
house_test_data['Bedroom'] = pd.Series(size_bed,index=house_test_data.index)

In [77]:
house_test_data

availability       size total_sqft  bath  balcony  price  Carpet  Area  \
0     Ready To Move      2 BHK       1225   2.0      2.0    NaN             0   
1     Ready To Move  9 Bedroom       2400   9.0      2.0    NaN             0   
2            18-Apr  4 Bedroom       1650   5.0      2.0    NaN             0   
3     Ready To Move      3 BHK       1322   3.0      1.0    NaN             0   
4     Ready To Move      2 BHK       1161   2.0      1.0    NaN             0   
...             ...        ...        ...   ...      ...    ...           ...   
1475  Ready To Move      2 BHK       1246   2.0      1.0    NaN             0   
1476  Ready To Move      3 BHK       1660   3.0      2.0    NaN             0   
1477  Ready To Move      2 BHK       1216   2.0      2.0    NaN             0   
1478         18-Feb      2 BHK        996   2.0      1.0    NaN             0   
1479  Ready To Move      2 BHK       1150   2.0      2.0    NaN             0   

      Plot  Area  Super built-up  Area  BHK  Bedroom  
0              0                     1    2        0  
1              1                     0    0        9  
2              1                     0    0        4  
3              0                     1    3        0  
4              0                     1    2        0  
...          ...                   ...  ...      ...  
1475           0                     1    2        0  
1476           0                     1    3        0  
1477           0                     1    2        0  
1478           0                     0    2        0  
1479           0                     0    2        0  

[1480 rows x 11 columns]

### DATE TIME:

In [78]:
avail_date = []
for item in house_test_data['availability']:
    if(item[0].isdigit()):
        item = str(datetime.today().date().year) + '-' + item
        avail_date.append((datetime.strptime(item , '%Y-%d-%b')).date())
    else:
        avail_date.append(0)
days_till_available = []
for item in avail_date:
    if(item != 0):
        n_days = item - datetime.today().date()
        if(n_days.days > 0):
            days_till_available.append(n_days.days)
        else:
            days_till_available.append(n_days.days + 365)
    else:
        days_till_available.append(0)

In [79]:
house_test_data['days_till_available'] = pd.Series(days_till_available , index=house_test_data.index)
house_test_data.head()

availability       size total_sqft  bath  balcony  price  Carpet  Area  \
0  Ready To Move      2 BHK       1225   2.0      2.0    NaN             0   
1  Ready To Move  9 Bedroom       2400   9.0      2.0    NaN             0   
2         18-Apr  4 Bedroom       1650   5.0      2.0    NaN             0   
3  Ready To Move      3 BHK       1322   3.0      1.0    NaN             0   
4  Ready To Move      2 BHK       1161   2.0      1.0    NaN             0   

   Plot  Area  Super built-up  Area  BHK  Bedroom  days_till_available  
0           0                     1    2        0                    0  
1           1                     0    0        9                    0  
2           1                     0    0        4                   64  
3           0                     1    3        0                    0  
4           0                     1    2        0                    0

In [80]:
house_test_data.drop(["size","availability"],axis=1,inplace=True)

In [81]:
house_test_data['balcony'].fillna(value=house_test_data['balcony'].mean(),inplace=True)
house_test_data['bath'].fillna(value=house_test_data['bath'].mean(),inplace=True)

In [82]:
house_test_data.isnull().sum()

total_sqft                 0
bath                       0
balcony                    0
price                   1480
Carpet  Area               0
Plot  Area                 0
Super built-up  Area       0
BHK                        0
Bedroom                    0
days_till_available        0
dtype: int64

In [83]:
house_test_data.drop(["price"],axis=1,inplace=True)
house_test_data.head()

total_sqft  bath  balcony  Carpet  Area  Plot  Area  Super built-up  Area  \
0       1225   2.0      2.0             0           0                     1   
1       2400   9.0      2.0             0           1                     0   
2       1650   5.0      2.0             0           1                     0   
3       1322   3.0      1.0             0           0                     1   
4       1161   2.0      1.0             0           0                     1   

   BHK  Bedroom  days_till_available  
0    2        0                    0  
1    0        9                    0  
2    0        4                   64  
3    3        0                    0  
4    2        0                    0

In [84]:
TOTAL_sqft = []
for item in house_test_data['total_sqft']:
    if('-' in str(item)):
        temp = item.split()
        temp1 = (float(temp[0]) + float(temp[2]))/2
        TOTAL_sqft.append(temp1)
    elif(item.endswith("Sq. Meter")):
        temp = float(item[:-9])
        TOTAL_sqft.append(temp * 10.7639)    
    elif(item.endswith("Perch")):
        temp = float(item[:-5])
        TOTAL_sqft.append(temp * 272.25)
    elif(item.endswith("Acres")):
        temp = float(item[:-5])
        TOTAL_sqft.append(temp * 43560)
    elif(item.endswith("Sq. Yards")):
        temp = float(item[:-9])
        TOTAL_sqft.append(temp * 9)
    elif(item.endswith("Cents")):
        temp = float(item[:-5])
        TOTAL_sqft.append(temp * 435.6)
    elif(item.endswith("Guntha")):
        temp = float(item[:-6])
        TOTAL_sqft.append(temp * 1089)
    elif(item.endswith("Grounds")):
        temp = float(item[:-7])
        TOTAL_sqft.append(temp * 2400)
    else:
        TOTAL_sqft.append(float(item))

house_test_data['TOTAL_sqft'] = pd.Series(TOTAL_sqft, index=house_test_data.index)
house_test_data.drop(columns='total_sqft',inplace=True)
house_test_data.head()

bath  balcony  Carpet  Area  Plot  Area  Super built-up  Area  BHK  \
0   2.0      2.0             0           0                     1    2   
1   9.0      2.0             0           1                     0    0   
2   5.0      2.0             0           1                     0    0   
3   3.0      1.0             0           0                     1    3   
4   2.0      1.0             0           0                     1    2   

   Bedroom  days_till_available  TOTAL_sqft  
0        0                    0      1225.0  
1        9                    0      2400.0  
2        4                   64      1650.0  
3        0                    0      1322.0  
4        0                    0      1161.0

In [85]:
len(house_test_data)

1480

In [86]:
len(house_data)

13320

## TRAIN-TEST SPLIT ON TRAIN DATA:

In [87]:
from sklearn.model_selection import train_test_split
X=house_data.drop('price',axis=1)
y=house_data['price']

In [89]:
y

0         39.07
1        120.00
2         62.00
3         95.00
4         51.00
          ...  
13315    231.00
13316    400.00
13317     60.00
13318    488.00
13319     17.00
Name: price, Length: 13320, dtype: float64

In [90]:
X_train,X_test,y_train,y_test = train_test_split(X,y,train_size=0.9) #90:10,80:20

In [94]:
len(y_test)

1332

### SVM REGRESSION ALGORITHM:

In [95]:
from sklearn.svm import SVR #best algo 
model = SVR()
model.fit(X_train,y_train)

SVR()

In [96]:
prediction = model.predict(X_test)

In [100]:
X_test

bath   balcony  Carpet  Area  Plot  Area  Super built-up  Area  BHK  \
9032    3.0  2.000000             0           0                     1    3   
3210    3.0  2.000000             0           0                     1    3   
4832    1.0  1.000000             0           0                     1    1   
7245    2.0  1.000000             0           0                     1    2   
1984    4.0  2.000000             0           1                     0    0   
...     ...       ...           ...         ...                   ...  ...   
7829    2.0  2.000000             0           0                     1    2   
4764    2.0  1.000000             0           0                     1    2   
3114    3.0  1.584376             0           0                     0    3   
10156   4.0  2.000000             0           0                     1    4   
5818    3.0  3.000000             0           0                     0    0   

       Bedroom  days_till_available  TOTAL_sqft  
9032         0                    0      1756.0  
3210         0                    8      2265.0  
4832         0                    0       780.0  
7245         0                    0       883.0  
1984         4                    0      1080.0  
...        ...                  ...         ...  
7829         0                    0      1222.0  
4764         0                    0      1216.0  
3114         0                    0      2820.0  
10156        0                  308      4395.0  
5818         4                    0      3532.0  

[1332 rows x 9 columns]

In [102]:
print(prediction)

[ 82.10853179  94.77987125  58.03811127 ... 108.60573923 147.14117463
 126.2311935 ]


In [101]:
y_test

9032     307.0
3210     136.0
4832      55.0
7245      37.0
1984     250.0
         ...  
7829      80.0
4764      84.0
3114     285.0
10156    242.0
5818     170.0
Name: price, Length: 1332, dtype: float64

In [104]:
model.score(X_test,y_test)

0.1306871302938475

In [105]:
from sklearn.model_selection import GridSearchCV  #improves accuracy

In [106]:
param_grid = {'C':[0.1,1,10,100,1000],'gamma':[1,0.1,0.01,0.001,0.0001],}

In [107]:
grid = GridSearchCV(SVR(),param_grid,verbose=3)

In [ ]:
#ex
grid.fit(X_train,y_train)

Fitting 5 folds for each of 25 candidates, totalling 125 fits
[CV] C=0.1, gamma=1 ..................................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ..................... C=0.1, gamma=1, score=-0.082, total=  10.7s
[CV] C=0.1, gamma=1 ..................................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   10.6s remaining:    0.0s


[CV] ..................... C=0.1, gamma=1, score=-0.068, total=  11.1s
[CV] C=0.1, gamma=1 ..................................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:   21.7s remaining:    0.0s


[CV] ..................... C=0.1, gamma=1, score=-0.069, total=  10.7s
[CV] C=0.1, gamma=1 ..................................................
[CV] ..................... C=0.1, gamma=1, score=-0.070, total=  10.6s
[CV] C=0.1, gamma=1 ..................................................
[CV] ..................... C=0.1, gamma=1, score=-0.065, total=  11.5s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ................... C=0.1, gamma=0.1, score=-0.073, total=  11.3s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ................... C=0.1, gamma=0.1, score=-0.063, total=  12.6s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ................... C=0.1, gamma=0.1, score=-0.063, total=  14.6s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ................... C=0.1, gamma=0.1, score=-0.064, total=  14.1s
[CV] C=0.1, gamma=0.1 ................................................
[CV] .

In [67]:
grid.best_params_

{'C': 1000, 'gamma': 0.0001}

In [68]:
grid.score(X_test,y_test)

0.48932216421392216

In [ ]:
grid.predict(X_test)

In [ ]:
y_test

In [69]:
prediction = grid.predict(house_test_data)

In [70]:
prediction

array([ 57.77780193, 349.07042026, 120.79222932, ...,  56.5957431 ,
        54.53523496,  55.78799633])